### ISSUE TITLE
#[derive(Into)]

### ISSUE THREAD

**Issue by dingfengquek** [JelteF/derive_more#13]
```rs
struct Oldtype;
struct Newtype(Oldtype);

impl From<Newtype> for Oldtype { ... } // #[derive(From)]
impl From<Oldtype> for Newtype { ... } // #[derive(Into)]

```

Code would be very similar to `#[derive(From)]`.
Although `#[derive(FromInverse)]` might be clearer, but longer.

**Comment by JelteF** [comment]
Good idea! I think whatever it would be named it it will be confusing though. I think the easiest way to avoid it would be to derive both when deriving From and possibly have long descriptive variants that only allow one way conversion.

**Comment by dingfengquek** [comment]
Some non-confusing very explicit names: `#[derive(FromSelfForInner)]` and `#[derive(FromInnerForSelf)]`, although they're somewhat long.

**Comment by dingfengquek** [comment]
I'm implementing for my own use because I want it badly. Would you like a PR for it in a few week's time?

**Comment by JelteF** [comment]
Yes, sounds great!

For structs it should be quite straightforward when looking at the opposite
From implementation and just reversing it should be fine. I would start
with those.

Enums are likely much more complicated. For them you probably have to
implement From for Option<InternalType>, since you don't know if it is the
correct variant. Also multiple variants could implement Into for the same
resulting type, they would need to be merged into a single Into impl. These
are the ones that are currently being ignored when deriving From, because
they would be ambiguous.

On Wed, Feb 8, 2017, 17:43 Dingfeng Quek <notifications@github.com> wrote:

I'm implementing for my own use because I want it badly. Would you like a
PR for it in a few week's time?

—
You are receiving this because you commented.


Reply to this email directly, view it on GitHub
<https://github.com/JelteF/derive_more/issues/13#issuecomment-278383912>,
or mute the thread

**Comment by JelteF** [comment]
Btw, I think calling it derive(Into) and having to derive it explicitly
would make the most sense after all. As long as we make the docs clear on
which does which it should be fine.

**Comment by JelteF** [comment]
If you just need derive(Into) for regular or tuple structs it's also fine if you just make a pull request for those btw. I can add the rest of the support later.

**Comment by JelteF** [comment]
Can you let me know if you're already  working on this. Otherwise I think I'll start on this in a couple of days.

**Comment by JelteF** [comment]
I just added an implementation of derive(Into) to the master branch (only for structs, not enums). 

I will close this issue for now. Could you test if it indeed covers your use case? Please reopen the issue if this is not the case.

**Comment by JelteF** [comment]
I just released 0.6.0, this also includes the generics fixes from before.
